libvast/include/vast/detail/adaptive_table_slice_builder_guards.hpp

```diff
@@ -0,0 +1,149 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2023 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#pragma once
+
+#include "vast/detail/series_builders.hpp"
+#include "vast/table_slice_builder.hpp"
+#include "vast/type.hpp"
+#include "vast/view.hpp"
+
+namespace vast::detail {
+
+class field_guard;
+
+/// @brief A view of a list column created within adaptive_table_slice_builder.
+/// Allows addition of new values into the list.
+class list_guard {
+private:
+  class list_record_guard {
+  public:
+    list_record_guard(concrete_series_builder<record_type>& builder,
+                      list_guard& parent, length_type starting_fields_length)
+      : builder_{builder},
+        parent_{parent},
+        starting_fields_length_{starting_fields_length} {
+    }
+
+    /// @brief Adds a field to a record nested inside a list.
+    /// @param name Field name.
+    /// @return Object used to append new values to a given field.
+    auto push_field(std::string_view name) -> field_guard;
+
+    ~list_record_guard() noexcept;
+
+  private:
+    concrete_series_builder<record_type>& builder_;
+    list_guard& parent_;
+    length_type starting_fields_length_ = 0u;
+  };
+
+public:
+  list_guard(concrete_series_builder<list_type>& root_list_builder,
+             list_guard* parent_list_guard, type value_type)
+
+    : builder_{root_list_builder},
+      parent{parent_list_guard},
+      value_type{std::move(value_type)} {
+  }
+
+  /// @brief Adds a value to a list. Use push_record and push_list to add record
+  /// and list respectively.
+  /// @param view View of a value to be added.
+  template <class ViewType>
+    requires requires(ViewType x) { materialize(x); }
+  auto add(ViewType view) -> void {
+    using view_value_type = decltype(materialize(std::declval<ViewType>()));
+    using vast_type = typename view_trait<view_value_type>::vast_type;
```

Comment:
The add is supposed to support everything but list, record and map (soon to be removed).
Adding support for accepting view<record> or view<list> can be done when needed. Currently i don't see a reason for it.
The implementation would just use push_record/push_list internally anyways i guess.

I actually see a need for add(const view&) in the json_parser.
 
 Do you see any use case for auto add(const type& t, const view& x) ?
 I think from the on demand parsing PoV we won't be using it. Any other ideas where it could be useful?

Reply:
Then we don't really need a type parameter. This is how part of the json parser looks right now
```
case simdjson::ondemand::number_type::floating_point_number: {
      auto result = val.get_double();
      if (result.error())
        return result.error();
      pusher.add(result.value_unsafe());
      return {};
    }
    case simdjson::ondemand::number_type::signed_integer: {
      auto result = val.get_int64();
      if (result.error())
        return result.error();
      pusher.add(result.value_unsafe());
      return {};
    }
    case simdjson::ondemand::number_type::unsigned_integer: {
      auto result = val.get_uint64();
      if (result.error())
        return result.error();
      pusher.add(result.value_unsafe());
      return {};
    }
```

The auto result variable in each switch case has the correct type. E.g double/int64_t uint64_t.
We can just use the currently implemented add to have the overload at the compile time.

Reply:
Right, for the auto-inference case that works out. Maybe it's fine then to change this afterwards.

I would still want to restrict the `view_traits<T>::vast_type` a bit so it only handles basic types. You could consider creating a list of views from the list of basic types, and then creating a trait that maps from view to basic type.

E.g.,

```c++
template <class T>
struct type_from_data
  : caf::detail::tl_at<
      concrete_types,
      caf::detail::tl_index_of<
        caf::detail::tl_map_t<concrete_types, type_to_data>, T>::value> {
};

template <class T>
using type_from_data_t = typename type_from_data<T>::type;

// Not sure if view_trait does the correct mapping, but you get the idea.
template <class T>
struct type_from_view
  : caf::detail::tl_at<
      concrete_types,
      caf::detail::tl_index_of<
        caf::detail::tl_map_t<
         caf::detail::tl_map_t<concrete_types, type_to_data>,
         view_trait>, T>::value> {
};

template <class T>
using type_from_view_t = typename type_from_view<T>::type;

Reply:
I have just used one of your implementation. Or do You want me to constraint it to basic types? I guess i could add req
If we ever introduce a type i'd like this to work automatically.
It could be done by making two type_lists

basic_types and complex_types.
Make concrete_types to be a sum of basic_types and complex_types. Then i could implement basic_type_from_data with in a similar manner. Or maybe just a static_assert that the type is a basic_type in the methods?